python/src/spark_rapids_ml/knn.py

```diff
@@ -590,32 +701,534 @@ async def do_allGather() -> List[str]:
+    phase, a query will be partitioned into a number of closest centers, and probe all the items associated with those centers. In
+    the end the top k closest items will be returned as the approximate nearest neighbors.
+
+    The current implementation build kmeans index independently on every GPU with its part of item vectors. Queries will be broadcast
+    to all GPUs, then every query probes closest centers on individual index. Local topk results will be aggregated to obtain
+    global topk ANNs.
+
+
+    Parameters
+    ----------
+    k: int (default = 5)
+        the default number of approximate nearest neighbors to retrieve for each query.
+
+    algorithm: str (default = 'ivfflat')
+        the algorithm parameter to be passed into cuML. It currently must be 'ivfflat'. Other algorithms such as
+        'ivfpq' are expected to be supported later.
+
+    algoParams: Optional[Dict[str, Any]] (default = None)
+        if set, algoParam is used to configure the algorithm.
+        When algorithm is 'ivfflat':
+            *nlist: (int) number of kmeans clusters to partition the dataframe into.
+            *nprobe: (int) number of closest clusters to probe for topk ANNs.
+
+    inputCol: str or List[str]
+        The feature column names, spark-rapids-ml supports vector, array and columnar as the input.\n
+            * When the value is a string, the feature columns must be assembled into 1 column with vector or array type.
+            * When the value is a list of strings, the feature columns must be numeric types.
+
+    idCol: str
+        the name of the column in a dataframe that uniquely identifies each vector. idCol should be set
+        if such a column exists in the dataframe. If idCol is not set, a column with the name `unique_id`
+        will be automatically added to the dataframe and used as unique identifier for each vector.
+
+    verbose:
+    Logging level.
+            * ``0`` - Disables all log messages.
+            * ``1`` - Enables only critical messages.
+            * ``2`` - Enables all messages up to and including errors.
+            * ``3`` - Enables all messages up to and including warnings.
+            * ``4 or False`` - Enables all messages up to and including information messages.
+            * ``5 or True`` - Enables all messages up to and including debug messages.
+            * ``6`` - Enables all messages up to and including trace messages.
+
+    Examples
+    --------
+    >>> from spark_rapids_ml.knn import ApproximateNearestNeighbors
+    >>> data = [(0, [0.0, 0.0]),
+    ...         (1, [1.0, 1.0]),
+    ...         (2, [2.0, 2.0]),
+    ...         (3, [30.0, 30.0]),
+    ...         (4, [40.0, 40.0]),
+    ...         (5, [50.0, 50.0]),]
+    >>> data_df = spark.createDataFrame(data, schema="id int, features array<float>")
+    >>> data_df = data_df.repartition(2) # ensure each partition having more data vectors than the 'nlist' of 'ivfflat'
```

Comment:
Why 2?  Isn't this done internal to repartition to number of gpus?

Reply:
by default spark shell master uses 'local[*]' and the data will be partitioned to num_cores partitions. 
With this toy dataset on my workstation, a partition has 1 or 0 vectors. 
Cuml requires the number of vectors to be smaller than the 'nlist' argument (i.e. 2 in our docstring).